internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,558 @@
+	lastBlockHash := &msg.InvList[lastBlock].Hash
+	_, err := p.headersService.GetHeightByHash(lastBlockHash)
+	if err == nil {
+		p.log.Info().Msgf("blocks from inv msg from peer %s already existsing in db", p)
+		return
+	}
+
+	p.log.Info().Msgf("requesting new headers from peer %s", p)
+	err = p.writeGetHeadersMsg(lastBlockHash)
+	if err != nil {
+		p.log.Error().Msgf("error while requesting new headers from peer %s, reason: %v", p, err)
+	}
+}
+
+func (p *Peer) handleHeadersMsg(msg *wire.MsgHeaders) {
+	receivedCheckpoint := false
+	lastHeight := int32(0)
+	headersReceived := 0
+
+	for _, header := range msg.Headers {
+		h, addErr := p.chainService.Add(domains.BlockHeaderSource(*header))
+
+		if service.HeaderAlreadyExists.Is(addErr) {
+			continue
+		}
+
+		if service.BlockRejected.Is(addErr) {
+			// TODO: ban peer
+			p.log.Error().Msgf("received rejected header %v from peer %s", h, p)
+			_ = p.Disconnect()
+			return
+		}
+
+		if service.HeaderSaveFail.Is(addErr) {
+			p.log.Error().Msgf("couldn't save header %v in database, because of %+v", h, addErr)
+			continue
+		}
+
+		if service.HeaderCreationFail.Is(addErr) {
+			p.log.Error().Msgf("couldn't create header from %v because of error %+v", header, addErr)
+			continue
+		}
+
+		if service.ChainUpdateFail.Is(addErr) {
+			p.log.Error().Msgf("when adding header %v couldn't update chains state because of error %+v", header, addErr)
+			continue
+		}
+
+		var err error
+		receivedCheckpoint, err = p.verifyCheckpointReached(h, receivedCheckpoint)
+		if err != nil {
+			// TODO: ban peer or lower peer sync score
+			_ = p.Disconnect()
+			return
+		}
+
+		lastHeight = h.Height
+		headersReceived += 1
+	}
+
+	if headersReceived == 0 {
+		p.log.Debug().Msgf("received only existing headers from peer: %s", p)
+		return
+	}
+
+	p.log.Info().Msgf(
+		"successfully received %d headers from peer %s, up to height %d",
+		headersReceived, p, lastHeight,
+	)
+
+	if receivedCheckpoint {
+		p.nextCheckpoint = findNextHeaderCheckpoint(p.checkpoints, p.nextCheckpoint.Height)
+		if p.nextCheckpoint == nil {
+			p.synced = true
+		}
+	}
+
+	if p.nextCheckpoint != nil {
+		p.log.Info().Msgf(
+			"requesting next headers batch from peer %s, height range %d - %d",
+			p, lastHeight, p.nextCheckpoint.Height,
+		)
+		err := p.writeGetHeadersMsg(p.nextCheckpoint.Hash)
+		if err != nil {
+			p.log.Error().Msgf("error requesting headers from peer %s, reason: %v", p, err)
+		}
+		return
+	}
+
+	p.log.Info().Msgf(
+		"checkpoints synced, requesting headers from height %d up to end of chain (zero hash) from peer %s",
+		lastHeight, p,
+	)
+	err := p.writeGetHeadersMsg(&zeroHash)
+	if err != nil {
+		p.log.Error().Msgf("error requesting headers from peer %s, reason: %v", p, err)
+	}
+}
+
+func (p *Peer) verifyCheckpointReached(h *domains.BlockHeader, receivedCheckpoint bool) (bool, error) {
+	if p.nextCheckpoint != nil && h.Height == p.nextCheckpoint.Height {
+		if h.Hash == *p.nextCheckpoint.Hash {
```

Comment:
Are you sure you won't get the `null pointer panic` here?

Reply:
Safety check added :)